performance_manager/lib/s3_utils.py

```diff
@@ -5,30 +5,72 @@
 import pandas
 import pyarrow.parquet as pq
 from pyarrow import fs
+import pyarrow
 
 
-def read_parquet(
+def _get_pyarrow_table(
     filename: Union[str, List[str]],
-    columns: Union[List[str], slice] = slice(None),
     filters: Optional[Union[Sequence[Tuple], Sequence[List[Tuple]]]] = None,
-) -> pandas.core.frame.DataFrame:
+) -> pyarrow.Table:
     """
-    read parquet file or files from s3 and return it as a pandas dataframe
+    internal function to get pyarrow table from parquet file(s)
     """
     active_fs = fs.S3FileSystem()
     if isinstance(filename, list):
         to_load = [f.replace("s3://", "") for f in filename]
     else:
         to_load = [filename.replace("s3://", "")]
 
+    # using `read_pandas` because `read` with "columns" parameter results in
+    # much slower file downloads for some reason...
+    return pq.ParquetDataset(
+        to_load, filesystem=active_fs, filters=filters
+    ).read_pandas()
+
+
+def read_parquet(
+    filename: Union[str, List[str]],
+    columns: Union[List[str], slice] = slice(None),
+    filters: Optional[Union[Sequence[Tuple], Sequence[List[Tuple]]]] = None,
+) -> pandas.core.frame.DataFrame:
+    """
+    read parquet file or files from s3 and return it as a pandas dataframe
+    """
     return (
-        pq.ParquetDataset(to_load, filesystem=active_fs, filters=filters)
-        .read_pandas()
-        .to_pandas()
+        _get_pyarrow_table(filename, filters)
+        .to_pandas(self_destruct=True)
```

Comment:
self destruct deletes the pyarrow table?

Reply:
It's supposed to, but it's marked as an experimental feature. I didn't really notice any significant memory savings from its use, but I figured it couldn't hurt.